File: app/src/main/java/ru/iandreyshev/parserrss/models/async/GetArticleImageTask.java
package ru.iandreyshev.parserrss.models.async;

import android.graphics.Bitmap;
import android.support.annotation.Nullable;

import ru.iandreyshev.parserrss.app.App;
import ru.iandreyshev.parserrss.app.Utils;
import ru.iandreyshev.parserrss.models.imageProps.IImageProps;
import ru.iandreyshev.parserrss.models.repository.Article;
import ru.iandreyshev.parserrss.models.web.HttpRequestHandler;

public final class GetArticleImageTask extends Task<Long, Void, Bitmap> {
    private static final int MAX_BYTES_COUNT = 1048576; // 1MB

    private final long mArticleId;
    private final IImageProps mImageProps;

    public static void execute(long articleId, final ITaskListener<Bitmap> listener, final IImageProps props) {
        new GetArticleImageTask(articleId, listener, props)
                .executeOnExecutor(TaskExecutor.getMultiThreadPool());
    }

    @Nullable
    @Override
    protected Bitmap doInBackground(final Long... articles) {
        Bitmap imageBitmap;
        final Article article = App.getDatabase().getArticleById(mArticleId);

        if (article == null) {
            return null;
        } else if (article.getImage() != null) {
            return mImageProps.configure(article.getImage());
        }

        final HttpRequestHandler mRequestHandler = new HttpRequestHandler(article.getImageUrl() == null ? "" : article.getImageUrl());
        mRequestHandler.setMaxContentBytes(MAX_BYTES_COUNT);

        final HttpRequestHandler.State requestResult = mRequestHandler.sendGet();

        final byte[] imageBytes = mRequestHandler.getBody();
        imageBitmap = Utils.toBitmap(imageBytes);

        if (requestResult == HttpRequestHandler.State.Success && imageBitmap != null) {
            App.getDatabase().updateArticleImage(mArticleId, imageBitmap);
            imageBitmap = mImageProps.configure(imageBitmap);
        }

        return imageBitmap;
    }

    private GetArticleImageTask(long articleId, final ITaskListener<Bitmap> listener, final IImageProps props) {
        super(listener);
        mArticleId = articleId;
        mImageProps = props;
    }
}

File: app/src/main/java/ru/iandreyshev/parserrss/models/async/GetArticleImageTask.kt
package ru.iandreyshev.parserrss.models.async

import android.graphics.Bitmap

import ru.iandreyshev.parserrss.app.App
import ru.iandreyshev.parserrss.models.extention.bitmap
import ru.iandreyshev.parserrss.models.imageProps.IImageProps
import ru.iandreyshev.parserrss.models.web.HttpRequestHandler

class GetArticleImageTask private constructor(
        private val _articleId: Long,
        listener: ITaskListener<Long, Void, Bitmap>,
        private val _imageProps: IImageProps) : Task<Long, Void, Bitmap>(listener) {

    companion object {
        private const val MAX_BYTES_COUNT = 1048576L // 1MB

        fun execute(id: Long, listener: ITaskListener<Long, Void, Bitmap>, props: IImageProps) {
            GetArticleImageTask(id, listener, props).executeOnExecutor(Task.EXECUTOR)
        }
    }

    override fun doInBackground(vararg articles: Long?): Bitmap? {
        val article = App.database.getArticleById(_articleId)
        val articleImage = App.database.getArticleImageByArticle(_articleId)

        articleImage?.bitmap?.let { return _imageProps.configureToView(it) }

        val imageUrl = article?.imageUrl ?: return null
        val mRequestHandler = HttpRequestHandler(imageUrl)
        mRequestHandler.maxContentBytes = MAX_BYTES_COUNT

        val requestResult = mRequestHandler.sendGet()
        val imageBytes = mRequestHandler.body
        var imageBitmap = imageBytes?.bitmap

        if (requestResult === HttpRequestHandler.State.Success && imageBitmap != null) {
            App.database.putArticleImage(_articleId, _imageProps.configureToMemory(imageBitmap))
            imageBitmap = _imageProps.configureToView(imageBitmap)
        }

        return imageBitmap
    }

    interface IEventListener : ITaskListener<Long, Void, Bitmap>
}
